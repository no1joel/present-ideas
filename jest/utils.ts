import {
  render as testingLibraryRender,
  RenderOptions,
} from "@testing-library/vue";
import linkify from "vue-linkify";
import { createMemoryHistory, createRouter } from "vue-router";
import routes from "../present_ideas/present_ideas/static/router/routes";
import getStore from "../present_ideas/present_ideas/static/store";

const mockRouter = createRouter({ routes, history: createMemoryHistory() });

export const render = async (component, options = {}) => {
  const store = getStore();
  const defaultRenderOptions: RenderOptions = {
    global: {
      plugins: [store, mockRouter],
      directives: { linkified: linkify },
    },
  };
  const renderOptionsWithDefaults = { ...defaultRenderOptions, ...options };

  const result = testingLibraryRender(component, renderOptionsWithDefaults);

  return result;
};
