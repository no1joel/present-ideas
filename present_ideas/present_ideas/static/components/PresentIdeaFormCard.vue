<template>
  <div class="card">
    <LoadingIndicator v-if="saving" />
    <div class="card-header">
      <h5 class="card-title m-0" v-linkified>Add New!</h5>
    </div>
    <div class="card-body">
      <form class="form-horizontal">
        <div class="form-group">
          <label for="thing-input">Thing*</label>
          <input
            type="text"
            name="thing"
            id="thing-input"
            class="form-control"
            placeholder="Something I want..."
            v-model="name"
          />
        </div>
        <div class="form-group">
          <label for="price-input">Price</label>
          <input
            type="text"
            name="price"
            id="price-input"
            class="form-control"
            placeholder="Enter a price or a range, e.g. £10-£20"
            v-model="price"
          />
        </div>
        <div class="form-group">
          <label for="notes-input">Notes</label>
          <textarea
            name="notes"
            id="notes-input"
            class="form-control"
            placeholder="Enter a link to a specific thing, a description of what you want, why you want it, anything!"
            v-model="notes"
          ></textarea>
        </div>
        <button
          type="button"
          class="btn btn-success btn-block"
          v-bind:disabled="addDisabled"
          v-on:click="add"
        >
          Add
        </button>
      </form>
    </div>
  </div>
</template>

<script>
import LoadingIndicator from "./LoadingIndicator.vue";

export default {
  components: { LoadingIndicator },
  props: {
    user: String,
  },
  data() {
    return {
      name: "",
      price: "",
      notes: "",
      saving: false,
    };
  },
  methods: {
    async add() {
      this.saving = true;
      await fetch("/api/add_idea/", {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          user: this.user,
          thing: this.name,
          price: this.price,
          notes: this.notes,
        }),
      });

      this.name = "";
      this.price = "";
      this.notes = "";

      this.saving = false;
      this.$emit("added");
    },
  },
  computed: {
    addDisabled() {
      return this.saving || !this.name;
    },
  },
};
</script>