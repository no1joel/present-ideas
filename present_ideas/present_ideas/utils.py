from typing import Iterable, List

from sheets.spreadsheet import (
    add_row,
    delete_row,
    get_rows,
    get_worksheets,
    set_cell_value,
)


def check_name(name: str) -> bool:
    """Check if the provided name is a sheet."""

    worksheet_names = get_all_names()
    return name.lower() in worksheet_names


def get_all_names() -> Iterable[str]:
    """Return people's names."""

    return (
        worksheet.title.lower()
        for worksheet in get_worksheets()
        if "template" not in worksheet.title.lower()
    )


def get_present_ideas(their_name: str) -> List[dict]:
    # TODO: Embed urls
    _header, *rows = get_rows(their_name)
    presents = [{**row, "index": index} for index, row in enumerate(rows)]
    return presents


def get_row_index(thing_index: int) -> int:
    # TODO: Why is it plus 3?!
    return thing_index + 3
