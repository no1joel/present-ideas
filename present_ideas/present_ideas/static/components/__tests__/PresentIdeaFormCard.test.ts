import { render } from "../../pages/__tests__/utils";
import PresentIdeaFormCard from "../PresentIdeaFormCard.vue";

describe("PresentIdeaFormCard", () => {
  it("matches snapshot", () => {
    expect(render(PresentIdeaFormCard).container).toMatchSnapshot();
  });
  it.todo("posts data to correct endpoint");
  it.todo("disables while saving and enables after");
});
