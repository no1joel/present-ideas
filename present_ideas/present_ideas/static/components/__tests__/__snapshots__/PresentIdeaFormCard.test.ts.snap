// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PresentIdeaCard matches snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <!---->
     
    <div
      class="card-header"
    >
      <h5
        class="card-title m-0"
      >
        
      
    
      </h5>
       
      <small>
        <!---->
         
        <!---->
      </small>
       
      <p
        class="card-text text-right mt-1"
      >
        <span>
          <transition-stub
            name=""
          >
            <span
              class=""
              style="display: none;"
            >
              <div
                class="popper popover"
              >
                <div
                  class="popover-header"
                >
                  Really delete?
                </div>
                 
                <div
                  class="popover-body"
                >
                  <button
                    class="btn btn-block btn-danger"
                    type="button"
                  >
                    
              Yep!
            
                  </button>
                </div>
              </div>
               
            </span>
          </transition-stub>
           
          <button
            class="btn btn-sm btn-outline-danger"
            type="button"
          >
            
          Delete
        
          </button>
        </span>
      </p>
    </div>
     
    <!---->
     
    <!---->
  </div>
</div>
`;

exports[`PresentIdeaFormCard matches snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <!---->
     
    <div
      class="card-header"
    >
      <h5
        class="card-title m-0"
      >
        Add New!
      </h5>
       
      <!---->
    </div>
     
    <div
      class="card-body"
    >
      <form
        class="form-horizontal"
      >
        <div
          class="form-group"
        >
          <label
            for="thing-input"
          >
            Thing*
          </label>
           
          <input
            class="form-control"
            id="thing-input"
            name="thing"
            placeholder="Something I want..."
            type="text"
          />
        </div>
         
        <div
          class="form-group"
        >
          <label
            for="price-input"
          >
            Price
          </label>
           
          <input
            class="form-control"
            id="price-input"
            name="price"
            placeholder="Enter a price or a range, e.g. £10-£20"
            type="text"
          />
        </div>
         
        <div
          class="form-group"
        >
          <label
            for="notes-input"
          >
            Notes
          </label>
           
          <textarea
            class="form-control"
            id="notes-input"
            name="notes"
            placeholder="Enter a link to a specific thing, a description of what you want, why you want it, anything!"
          />
        </div>
         
        <button
          class="btn btn-success btn-block"
          disabled="disabled"
          type="button"
        >
          
        Add
      
        </button>
      </form>
    </div>
  </div>
</div>
`;
